### PR TITLE
Clean up directories and files left behind by Toil (Resolves #2275)

### LINKED ISSUE #2275: Toil tests (?) make lots of temporary files, don't clean them up
Apparently I own 2000 things in the Courtyard temp directory, most of which are Toil-related temp directories.

Some of them look like this:

```
drwxr-xr-x     2 anovak   giuser    6 Jun 25 10:24 toil-35cf0ce4-b7ff-4c03-b370-d84e9d63ff3d-8e15d9a1-aa58-42e9-92b5-05ceb0c653f8
```

Some of them look like this:

```
drwx------     3 anovak   giuser  107 Jun 19 13:33 toil-test-toil.test.batchSystems.batchSystemTest.SingleMachineBatchSystemJobTest-testJobConcurrency-testFiles-EQNFVc
```

And some of them have no particular name structure but have listings that look like this:

```
[anovak@courtyard haplotype-aware-mapping-paper]$ ls -lah /data/tmp/tmpyBsbxl
total 120K
drwx------    4 anovak giuser 150 Jun 22 16:20 .
drwxrwxrwt 1764 root   root   76K Jul  3 15:50 ..
-rw-r--r--    1 anovak giuser   9 Jun 22 16:20 803d3b2fa54eca5fc5b36a02ebf169790e7f22d8
-rw-r--r--    1 anovak giuser   0 Jun 22 16:20 a7cb310b906054b3b77f2bc87ba58572acf9d9fc
drwx------    2 anovak giuser   6 Jun 22 16:20 out_tmpdira0KoLr
drwx------    2 anovak giuser 102 Jun 22 16:20 out_tmpdirwMFHiJ
```

Or this:

```
[anovak@courtyard haplotype-aware-mapping-paper]$ ls -lah /data/tmp/tmpoSgPZo
total 140K
drwxr-xr-x    3 anovak giuser  112 Jun 25 10:39 .
drwxrwxrwt 1764 root   root    76K Jul  3 15:50 ..
-rw-r--r--    1 anovak giuser 1.5K Jun 25 10:39 config.pickle
-rw-r--r--    1 anovak giuser 9.1K Jun 25 10:39 environment.pickle
-rw-r--r--    1 anovak giuser   82 Jun 25 10:39 rootJobReturnValue
-rw-r--r--    1 anovak giuser   13 Jun 25 10:39 rootJobStoreID
drwxr-xr-x    6 anovak giuser   42 Jun 25 10:39 tmp
```

I suspect these directories are a result of running the toil tests, and some of them may result from the tests failing or being killed.

The Toil tests should always clean up their temp files when they pass, and should also at least try to clean up their temp files when they fail or are killed.

### PR DESCRIPTION
**Pull request by bencvdb**
There are multiple places in Toil where files and directories are left behind.

This PR would clean up those files.

### CODE REVIEW

**Comment by DailyDreaming** [comment]
Failing tests:
http://jenkins.cgcloud.info/job/toil-pull-request-integration-tests/2634/
http://jenkins.cgcloud.info/job/toil-pull-requests/4347/

**Comment by DailyDreaming** [comment]
testGoogleMesos might be stocastic?
http://jenkins.cgcloud.info/job/toil-pull-requests/4350/testReport/junit/src.toil.test.sort.sortTest/SortTest/testGoogleMesos/
http://jenkins.cgcloud.info/job/toil-pull-request-integration-tests/2637/testReport/junit/src.toil.test.provisioners.aws.awsProvisionerTest/AWSAutoscaleTest/testSpotAutoScale/

Testing again, just in case.

**Comment by DailyDreaming** [comment]
Jenkins, test this please.

**Comment by bencvdb** [comment]
Failed Tests:
http://jenkins.cgcloud.info/job/toil-pull-requests/4351/testReport/junit/src.toil.test.src.fileStoreTest/CachingFileStoreTestWithGoogleJobStore/testAsyncWriteWithCaching/

http://jenkins.cgcloud.info/job/toil-pull-request-integration-tests/2638/testReport/junit/src.toil.test.provisioners.aws.awsProvisionerTest/AWSRestartTest/testAutoScaledCluster/

**Comment by bencvdb** [comment]
Jenkins, test this please.

**Comment by DailyDreaming** [comment]
Latest tests:
http://jenkins.cgcloud.info/job/toil-pull-request-integration-tests/2645/
http://jenkins.cgcloud.info/job/toil-pull-requests/4357/

File: src/toil/cwl/cwltoil.py [hunk 1]
```diff
@@ -987,6 +986,13 @@ def visitSteps(t, op):
             visitSteps(s.embedded_tool, op)
 
 
+def cleanTempDirs(job):
+    """Remove temporarly created directories."""
+    if job._succeeded:
+        for tempDir in job.openTempDirs:
+            shutil.rmtree(tempDir)
```
Comment (DailyDreaming) [inline src/toil/cwl/cwltoil.py]: You may wish to have an `if os.path.exists():` statement here, as well as resetting with: `job.openTempDirs = []` when it has finished deleting the directories, just to be safe.

Comment (bencvdb) [reply src/toil/cwl/cwltoil.py]: Good call.

File: src/toil/cwl/cwltoil.py [hunk 2]
```diff
@@ -542,9 +543,8 @@ def run(self, file_store):
         outdir = os.path.join(file_store.getLocalTempDir(), "out")
         os.mkdir(outdir)
         top_tmp_outdir = self.workdir or os.environ["TMPDIR"]
-        tmp_outdir_prefix = os.path.join(
-            _makeNestedTempDir(top=top_tmp_outdir, seed=outdir, levels=2),
-            "out_tmpdir")
+        tmp_outdir_prefix = os.path.join(_makeNestedTempDir(top=top_tmp_outdir, seed=outdir, levels=2), "out_tmpdir")
```
Comment (mr-c) [inline src/toil/cwl/cwltoil.py]: Please line-wrap

Comment (bencvdb) [reply src/toil/cwl/cwltoil.py]: Will do!

File: src/toil/cwl/cwltoil.py [hunk 3]
```diff
@@ -987,6 +986,14 @@ def visitSteps(t, op):
             visitSteps(s.embedded_tool, op)
 
 
+def cleanTempDirs(job):
+    """Remove temporarly created directories."""
+    if job is CWLJob and job._succeeded:  # Only CWLJobs have this attribute.
+        for tempDir in job.openTempDirs:
+            if os.path.exists(tempDir):
+                shutil.rmtree(tempDir)
+
```
Comment (DailyDreaming) [inline src/toil/cwl/cwltoil.py]: Remove `tempDir` from `job.openTempDirs` once it's deleted.

**Comment by DailyDreaming** [comment]
LGTM!

**Pull request merged**